### PR TITLE
GEODE-4559: pass the Cache to a Declarable

### PR DESCRIPTION
**Pull request by dschneider-pivotal**
The Declarable interface now has "initialize" on it and "init" has been deprecated.
The new initialize is also passed the Cache.
The AutoBalancer, which is a Declarable, uses the new initialize instead of calling getAnyInstance

### CODE REVIEW

File: geode-core/src/main/java/org/apache/geode/internal/cache/xmlcache/CacheCreation.java [hunk 1]
```diff
@@ -550,14 +577,31 @@ void create(InternalCache cache)
...
+      try {
+        declarable.setCache(cache);
+        declarable.init(properties);
+      } catch (Exception ex) {
+        throw new CacheXmlException(
+            "Exception while initializing an instance of " + declarable.getClass().getName(), ex);
+      }
+      this.declarablePropertiesMap.put(declarable, properties);
+    }
+    this.declarablePropertiesList.clear();
```
Comment (agingade) [inline geode-core/src/main/java/org/apache/geode/internal/cache/xmlcache/CacheCreation.java]: This should be in finally block?

File: geode-core/src/main/java/org/apache/geode/internal/DeployedJar.java [hunk 2]
```diff
@@ -298,8 +298,8 @@ boolean hasSameContentAs(final File stagedFile) {
         boolean registerUninitializedFunction = true;
         if (Declarable.class.isAssignableFrom(clazz)) {
           try {
-            final List<Properties> propertiesList = ((InternalCache) CacheFactory.getAnyInstance())
-                .getDeclarableProperties(clazz.getName());
+            InternalCache cache = (InternalCache) CacheFactory.getAnyInstance();
+            final List<Properties> propertiesList = cache.getDeclarableProperties(clazz.getName());
```
Comment (agingade) [inline geode-core/src/main/java/org/apache/geode/internal/DeployedJar.java]: propertiesList could be null.

Comment (dschneider-pivotal) [reply geode-core/src/main/java/org/apache/geode/internal/DeployedJar.java]: Checked this and it is never null

File: geode-core/src/main/java/org/apache/geode/internal/cache/xmlcache/CacheCreation.java [hunk 3]
```diff
@@ -550,14 +577,31 @@ void create(InternalCache cache)
     }
 
     cache.setBackupFiles(this.backups);
-    cache.addDeclarableProperties(this.declarablePropertiesMap);
-    runInitializer();
+
+    runInitializer(cache);
```
[the remark this answers is not in the document]
Comment (dschneider-pivotal) [reply geode-core/src/main/java/org/apache/geode/internal/cache/xmlcache/CacheCreation.java]: Fixed this. runInitializer now uses the cache. Added unit tests to verify

**Pull request merged**